sympy/solvers/solveset.py

```diff
@@ -3479,19 +3483,52 @@ def _separate_poly_nonpoly(system, symbols):
             polys_expr.append(poly.as_expr())
         else:
             nonpolys.append(eq)
-    return polys, polys_expr, nonpolys, denominators
+    return polys, polys_expr, nonpolys, denominators, unrad_changed
 # end of def _separate_poly_nonpoly()
 
 
+def _handle_poly(polys, symbols):
+    poly_sol = [{}]
+    # Compute a Groebner basis in lex order wrt the ordering given by
+    # symbols.
+    #
+    # XXX: Maybe it is possible to choose a better ordering of the symbols
+    # here like choose the symbol was highest (or lowest?) degree first...
+
+    basis = groebner(polys, symbols, polys=False)
+    # Does the polynomial system have a finite number of solutions?
+    if basis.is_zero_dimensional:
+
+        # Solve the zero-dimensional case using solve_poly_system if
+        # possible. Otherwise fall back on using substitution below.
+        #
+        # XXX: Why would solve_poly_system fail and how would subsitution
+        # do any better in that case?
+        try:
+            result = solve_poly_system(basis, *symbols)
+        except NotImplementedError:
+            # Use the Groebner basis instead of the original polys for
+            # further solving with substitution.
+            poly_eqs = list(basis)
+        else:
+            poly_sol = [dict(zip(symbols, res)) for res in result]
+            poly_eqs = []
```

Comment:
The complicated logic in this function can be simplified now that it is a function. It's just this:
```python
    if basis.is_zero_dimensional:
        try:
            ... = solve_poly_system(...)
        except NotImplementedError:
            pass
        else:
            return ...

    poly_eqs = list(basis)
    poly_sol = [{}]
    return poly_sol, poly_eqs
```
The questions that I left in the comments were things to think about e.g.:
```
        # XXX: Why would solve_poly_system fail and how would subsitution
        # do any better in that case?
```
Are there any test cases that follow this branch and is so what is the answer to the question?

Reply:
I don't think there are any tests that currently go to that branch. All tests in the solvers module pass even if that branch is removed. I tried following the method calls and there are a few places in `solve_generic` (which is called by `solve_poly_system`) where `NotImplementedError` is raised for the system being positive-dimensional, but that should not be a problem here since the system is guaranteed to be zero-dimensional. There could also be many other places where `NotImplementedError` is potentially raised, but it is not clear to me how substitution would help in such cases. 
Also, on master the comment under that branch says that `_handle_zero_dimensional`, which calls `solve_poly_system`, does not currently fail for any polynomial equation:
https://github.com/sympy/sympy/blob/0bc2587aef09ab150bd408b7cd9e08baf7b5fd9c/sympy/solvers/solveset.py#L3678-L3680
Considering this, can that branch be safely removed?